### PR TITLE
Enabled g_mem_set_vtable through 

### PR DESCRIPTION
**Pull request by Dunati**
For embedding and tracking down memory leaks it can be useful to replace malloc/free with custom versions. This is enabled through a configure option which defines G_OVERRIDABLE_ALLOCATORS. An example of usage is added to the embedded sample.

Also fixed is consistently using the gmem allocation methods in eglib.

### CODE REVIEW

File: eglib/configure.ac [hunk 1]
```diff
@@ -138,6 +138,17 @@ AC_CHECK_FUNCS(strlcpy stpcpy strtok_r rewinddir vasprintf)
 AC_CHECK_FUNCS(getrlimit)
 AC_CHECK_FUNCS(fork execv execve)
 
+AC_ARG_WITH([overidable-allocators], [  --with-overridable-allocators	allow g_*alloc/g_free to call custom allocators set via g_mem_set_vtable])
+
+if test x$with_overridable_allocators == xyes; then
+	ENABLE_OVERRIDABLE_ALLOCATORS="1"
```
Comment (kumpera) [inline eglib/configure.ac]: Use AC_DEFINE instead of setting a var here and checking it in config.h.in

File: eglib/configure.ac [hunk 2]
```diff
@@ -151,7 +162,7 @@ elif test x$target_ios = xno; then
 AC_CHECK_FUNCS(strndup getpwuid_r)
 fi
 
-AM_CONDITIONAL(NEED_VASPRINTF, test x$ac_cv_func_vasprintf = xno )
+AM_CONDITIONAL(NEED_VASPRINTF, test x$ac_cv_func_vasprintf = xno || test x$with_overridable_allocators == xyes)
```
Comment (kumpera) [inline eglib/configure.ac]: This change makes no sense, why would we force a vasprintf implementation on targets that have it?

File: eglib/src/gmisc-win32.c [hunk 3]
```diff
@@ -128,6 +128,9 @@ g_get_home_dir (void)
 		}
 	}
 
+	g_free (drive);
```
Comment (kumpera) [inline eglib/src/gmisc-win32.c]: Those 2 free calls are out of context here. Do we have a mem leak?

**Comment by kumpera** [comment]
Hi Dunati,

There's a lot of eglib cleanup in the patch that paves the work to help with the scenario you're trying to support.

My problem is that I don't like the public interface that glib offers.

Do you mind raising the design discussion for this on mono-dev?

File: eglib/src/glib.h [hunk 4]
```diff
@@ -154,19 +155,21 @@ gpointer g_try_realloc (gpointer obj, gsize size);
 #define g_alloca(size)		alloca (size)
 
 gpointer g_memdup (gconstpointer mem, guint byte_size);
-static inline gchar   *g_strdup (const gchar *str) { if (str) {return strdup (str);} return NULL; }
+static inline gchar   *g_strdup (const gchar *str) { if (str) { return (gchar*) g_memdup(str, (guint)strlen (str) + 1); } return NULL; }
```
Comment (txdv) [inline eglib/src/glib.h]: style nit: there should be a space after g_memdup, it is a function after all

File: eglib/configure.ac (hunk 2 above)
Comment (Dunati) [inline eglib/configure.ac]: The problem is that the system vasprintf allocates memory with malloc, which which is a problem if you then try to free that memory with a overridden g_free

File: eglib/src/gmisc-win32.c (hunk 3 above)
Comment (Dunati) [inline eglib/src/gmisc-win32.c]: yes, g_getenv returns a g_malloc'd string on win32

File: mono/utils/mono-publib.h [hunk 5]
```diff
@@ -69,6 +71,19 @@ typedef uint32_t	mono_unichar4;
...
+	void* (*realloc) (void* mem, size_t n_bytes);
+	void  (*free)    (void* mem);
+	void* (*calloc)  (size_t n_blocks, size_t n_block_bytes);
+} MonoAllocatorVTable;
+
+/* 
+ * eglib must be built with overridable allocator support (G_OVERRIDABLE_ALLOCATORS=1) 
+ * for mono_set_allocator_vtable to do anything.
+ */
+MONO_API void mono_set_allocator_vtable (MonoAllocatorVTable* vtable);
```
Comment (Dunati) [inline mono/utils/mono-publib.h]: Okay, I can pull it out. I will have to export g_mem_set_vtable in the .def files and that felt strange with no other eglib methods exported. It also requires digging into the headers a bit.

**Comment by Dunati** [comment]
I cleaned things up based on the the feedback so far. Who should I ping in monodev?

**Comment by Dunati** [comment]
I also have a similar change for the mono_valloc family of functions 
which I am probably just going to keep in our fork unless a design discussion comes up with a good solution for both sets of allocators.

https://github.com/Dunati/mono/blob/mono-4.4.0-linden/mono/utils/mono-mmap.h 
```
typedef struct _MonoAllocMethods
{
	int   (*mono_pagesize)   (void);
	void* (*mono_valloc)     (void *addr, size_t length, int flags);
	void* (*mono_valloc_aligned) (size_t length, size_t alignment, int flags);
	int   (*mono_vfree)      (void *addr, size_t length);
	int(*mono_mprotect)(void *addr, size_t length, int flags);
} MonoAllocMethods;

MONO_API void  mono_set_alloc_methods (MonoAllocMethods*);
```

**Comment by kumpera** [comment]
@Dunati My major pet peeve with the glib style of functions is that they give no room for the runtime to provide meaning to the allocations. @evincarofautumn has a PR with some nice bits for how to do it with the valloc family of functions.

**Comment by kumpera** [comment]
No one in special needs to be pinged on monodev. Let me bootstrap the conversation then.

**Comment by kumpera** [comment]
@Dunati After some discussion with the team, we'll extend the profiling interface to provide reporting instead of the alloc override.

One more thing. I don't see the value of making this optional. Why did you do it that way?

**Comment by Dunati** [comment]
With this patch there is an extra pointer dereference per call to fetch the allocation method. Running in a dll (as on windows) I don't think the compiler can optimize that away. I can make the patch much more simple if it is not optional.

Also, I have been hanging out in #monodev on irc.gnome.org is that the right place?

**Comment by kumpera** [comment]
We've being using gitter

**Comment by kumpera** [comment]
Manually merged this PR to fix the conflicts and add back the mono vtable part

**Pull request closed without merging**